### PR TITLE
fix: clean compaction output from scratchpad

### PR DESCRIPTION
**Pull request by Joe-Blount**
Helps: https://github.com/influxdata/idpe/issues/17702

The compactor repeatedly compacts partitions with a lot of files needing compacted.  Such partitions can continue being compacted until the partition timeout.  In one cluster with many long running partitions with large files, the compactor's scratchpad memory use grows until the timeout, at which point memory use drops significantly.

Compaction output files were not removed from the scratchpad during these many little compactions on a partition.  Its possible that the compaction output from one round may compaction input for a later round, so keeping them can have benefit.  But keeping all the compaction output files with the hope of some of them being reused appears to be the cause of the high memory use.  So this PR removes the compaction output files from the scratchpad after they are uploaded to objectstore.  

Additionally, the cleaning of compaction input files is moved up a few lines. This isn't critical to the defect fix, but since uploading to objectstore may take some time, we might as well free that memory a bit sooner.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
I may be missing something, but it seems to me like this PR effectively disables most of the benefit of the scratchpad 🤔 

If the goal of this PR is to measure the effect of the scratchpad in production, another way to do so might be to add a flag that would configure the compactor to use the `NoopScratchpad` ([source](https://github.com/influxdata/influxdb_iox/blob/jrb_46_compactor_scratchpad_cleaning/compactor/src/components/scratchpad/noop.rs#L32))

Then we could test the effect of enabling/disabling the scratchpad on one cluster before rolling it out to all clusters


BTW the CI test https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/34193/workflows/13bda8c4-c725-420b-9d84-60ebf1b1f5da/jobs/319337 is  failing:


Like this 
```
test test_shadow_mode has been running for over 60 seconds

Too long with no output (exceeded 10m0s): context deadline exceeded
```

When I ran that test locally like this (to see logging):

```shell
RUST_LOG=trace cargo test -p compactor --test integration -- --nocapture test_shadow_mode
```

I see this in the log:

```
2023-06-08T18:42:12.717992Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:12.920836Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:13.101572Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:13.565071Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:14.506969Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=1
2023-06-08T18:42:16.304770Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:17.304301Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:17.593388Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:18.423887Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:18.995020Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:19.514106Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=1
2023-06-08T18:42:20.759970Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=1
2023-06-08T18:42:21.922094Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=2
2023-06-08T18:42:24.784762Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=3
2023-06-08T18:42:28.218513Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=3
2023-06-08T18:42:32.116984Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:33.057600Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:33.253651Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:33.622373Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=0
2023-06-08T18:42:34.622072Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=1
2023-06-08T18:42:35.977371Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=3
2023-06-08T18:42:39.253872Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=6
2023-06-08T18:42:45.852301Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=14
test test_shadow_mode has been running for over 60 seconds
2023-06-08T18:43:00.123543Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=6
2023-06-08T18:43:06.137759Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=3
2023-06-08T18:43:09.402177Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=4
2023-06-08T18:43:13.713010Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=4
2023-06-08T18:43:18.168802Z  WARN backoff: request encountered non-fatal error - backing off error=Object at location 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet not found: No data in memory found. Location: 1/1/1/7602d9d8-2554-00b6-3cb5-385da8e5b888.parquet task_name="copy file" backoff_secs=6
```

Which seems related to this PR -- perhaps it is because shadow mode doesn't actually write anything to the object store 🤔

**Comment by Joe-Blount** [comment]
> I may be missing something, but it seems to me like this PR effectively disables most of the benefit of the scratchpad 

I think there's two categories of benefit to the scratchpad:
1) preventing the latency of objectstore access from impacting our compaction jobs (e.g. if DF jobs had to wait on moving objects to/from objectstore, would performance be bad, or more specifically would performance be bad relative to DF's memory usage?) - this benefit is unaffected by this PR.
2) reducing quantity of data transferred to/from objectstore by caching the output of compactions in anticipation it might be used in the next round of compaction.  You are correct that this benefit is removed by this PR.

Personally I'm not too concerned about negating the caching benefit, unless it is shown to cause a problem.  Partly that's because we can compensate by prefetching if necessary.  Party, that's because the opportunity for caching benefit fluctuates based on a lot of other decisions we make.  For example, if an external scheduler micro-manages the compaction jobs, giving a compactor only tiny work items, there is no opportunity for caching from one round of compaction to the next.   If our current performance relies on the caching benefit between compaction rounds, it would be good to know that before we make other compaction design decisions that might indirectly reduce the opportunity for such caching.

If it is worthwhile, could maybe get smart about cleaning these compaction output files from the scratchpad with some amount of lag (if we haven't reused it by <>, assume we won't and clean it), but I'd rather not go there.  For now I'd like to simply remove all of the compaction output from the scratchpad.  If that eliminates the memory climb I'm chasing, we'll have clear confirmation the problem isn't something else (like overlooked cleanup of compaction input in some cases).  And we might learn something about why we do or don't need the scratchpad.

> BTW the CI test is failing

:cry: yes, [I've been bitten by that before](https://influxdata.slack.com/archives/C04QYLRFQE5/p1682710388215929).
as I understand it, the intent of shadow mode is in conflict with the intent of this PR.  I'm looking at making the change in this PR conditional on `! shadow_mode`.

**Comment by alamb** [comment]
> as I understand it, the intent of shadow mode is in conflict with the intent of this PR. I'm looking at making the change in this PR conditional on ! shadow_mode.

This makes sense to me

> If it is worthwhile, could maybe get smart about cleaning these compaction output files from the scratchpad with some amount of lag (if we haven't reused it by <>, assume we won't and clean it), but I'd rather not go there. For now I'd like to simply remove all of the compaction output from the scratchpad. If that eliminates the memory climb I'm chasing, we'll have clear confirmation the problem isn't something else (like overlooked cleanup of compaction input in some cases). And we might learn something about why we do or don't need the scratchpad.

This also makes sense to me

I was mostly thinking that the scratchpad add some non trivial complexity (in my mind) as it adds a few layers of indirection between reading/writing parquet files and when they are actually read/written to object store.

So basically if we disable most of the benefit gained by the scratchpad I think we should also try and remove as much of (the now unnecessary) code as well

**Review by alamb — APPROVED** [review]

**Comment by Joe-Blount** [comment]
Since there's not urgency on this, I'll not merge it on Friday. I plan to merge this Monday

With this PR, I hope to see memory use decrease significantly, which will confirm our understanding of the cause of the memory use.
If performance also remains good with this PR merged, it confirms the caching-for-potential-future-reuse benefit from the scratchpad is not critical.
If both of those points work out as hoped, the next step is to remove the scratchpad and see if performance remains good.  We may find that the latency protection offered by the scratchpad is necessary for performance, if which case we may need to stick with what's in this PR.  All that should become clear shortly...

File: compactor/src/driver.rs [hunk 1]
```diff
@@ -345,9 +348,9 @@ async fn execute_branch(
         .map(ParquetFilePath::from)
         .collect();
 
-    // Now that we've uploaded them, compaction output files can be removed from the scratchpad.
+    // conditionally (if not shaddow mode) remove the newly created files from the scratchpad.
```
Comment (alamb) [inline compactor/src/driver.rs:351]: 👍  this makes sense to me

**Pull request merged**